beetsplug/musicbrainz.py

```diff
@@ -0,0 +1,919 @@
+# permit persons to whom the Software is furnished to do so, subject to
+# the following conditions:
+#
+# The above copyright notice and this permission notice shall be
+# included in all copies or substantial portions of the Software.
+
+"""Searches for albums in the MusicBrainz database."""
+
+from __future__ import annotations
+
+import re
+import traceback
+from collections import Counter
+from itertools import product
+from typing import TYPE_CHECKING, Any
+from urllib.parse import urljoin
+
+import musicbrainzngs
+
+import beets
+import beets.autotag.hooks
+from beets import config, plugins, util
+from beets.plugins import BeetsPlugin, MetadataSourcePlugin
+from beets.util.id_extractors import (
+    beatport_id_regex,
+    deezer_id_regex,
+    extract_discogs_id_regex,
+    spotify_id_regex,
+)
+
+if TYPE_CHECKING:
+    from collections.abc import Iterator
+
+    from beets.library import Item
+
+    from ._typing import JSONDict
+
+VARIOUS_ARTISTS_ID = "89ad4ac3-39f7-470e-963a-56509c546377"
+
+BASE_URL = "https://musicbrainz.org/"
+
+SKIPPED_TRACKS = ["[data track]"]
+
+FIELDS_TO_MB_KEYS = {
+    "catalognum": "catno",
+    "country": "country",
+    "label": "label",
+    "barcode": "barcode",
+    "media": "format",
+    "year": "date",
+}
+
+musicbrainzngs.set_useragent("beets", beets.__version__, "https://beets.io/")
+
+
+class MusicBrainzAPIError(util.HumanReadableError):
+    """An error while talking to MusicBrainz. The `query` field is the
+    parameter to the action and may have any type.
+    """
+
+    def __init__(self, reason, verb, query, tb=None):
+        self.query = query
+        if isinstance(reason, musicbrainzngs.WebServiceError):
+            reason = "MusicBrainz not reachable"
+        super().__init__(reason, verb, tb)
+
+    def get_message(self):
+        return "{} in {} with query {}".format(
+            self._reasonstr(), self.verb, repr(self.query)
+        )
+
+
+RELEASE_INCLUDES = list(
+    {
+        "artists",
+        "media",
+        "recordings",
+        "release-groups",
+        "labels",
+        "artist-credits",
+        "aliases",
+        "recording-level-rels",
+        "work-rels",
+        "work-level-rels",
+        "artist-rels",
+        "isrcs",
+        "url-rels",
+        "release-rels",
+        "tags",
+    }
+    & set(musicbrainzngs.VALID_INCLUDES["release"])
+)
+
+TRACK_INCLUDES = list(
+    {
+        "artists",
+        "aliases",
+        "isrcs",
+        "work-level-rels",
+        "artist-rels",
+    }
+    & set(musicbrainzngs.VALID_INCLUDES["recording"])
+)
+
+BROWSE_INCLUDES = [
+    "artist-credits",
+    "work-rels",
+    "artist-rels",
+    "recording-rels",
+    "release-rels",
+]
+if "work-level-rels" in musicbrainzngs.VALID_BROWSE_INCLUDES["recording"]:
+    BROWSE_INCLUDES.append("work-level-rels")
+BROWSE_CHUNKSIZE = 100
+BROWSE_MAXTRACKS = 500
+
+
+def _preferred_alias(aliases: list[JSONDict]):
```

Comment:
Can we get a return type hint here? Also, I think the docstring needs a bit of improvement.

Reply:
Note that in 4136d3e I only resolved typing errors, so the files that I adjusted are mostly 'clean' 😅 

As much as I'd like to add types here, I think I will leave this for a separate PR given that this one is already fairly large.

I've actually got a local branch where I manually defined `TypedDict`s for each of the resources used in here - maybe it's time to complete it